### ISSUE TITLE
Group/display large amount of overlays/symbols

### ISSUE THREAD

**Issue by maxBombrun** [openseadragon/openseadragon#1225]
Hello,

I would like to display a large number of elements, a bit like #857 .
I have, let say, 10 different svg symbols with various number of positions (can be up to 20000).
I found that "symbol", "use", and/or "markers" could help me.
Currently, I am using it like this, in javascript:

```
for (var idxSymbol in annotation_data.sets){
            for (var idxPoint in annotation_data.sets[idxSymbol].points){
                                svgImg= returnSvgSymbol(symbolName, symbol);
                                
                                OSViewer.addOverlay({
                                    id: svgImg,
                                    px: annotation_data.sets[idxSignal].points[idxPoint].x,
                                    py: annotation_data.sets[idxSignal].points[idxPoint].y,
                                    placement: 'CENTER',
                                    checkResize: false,
                                }); 
             }
}


function returnSvgSymbol(symbolName, symbolValue){
            var img = document.createElementNS("http://www.w3.org/2000/svg", "svg");
            var use = document.createElementNS('http://www.w3.org/2000/svg', 'use');
            
            var symbolType = (symbolValue.image.rect.y-1)%(tempSymb.length);
            var colorType = Math.floor((symbolValue.image.rect.y-1)/(tempSymb.length*10));
            var symb= "img/myicons.svg#".concat(tempSymb[symbolType]);

            use.setAttributeNS("http://www.w3.org/1999/xlink", "xlink:href", symb);
            use.setAttribute('style',"fill: transparent; stroke: ".concat(tempColor[colorType]).concat("; stroke-width:8%"));
            img.setAttribute('class', symbolName);
            img.setAttribute('width', '15');
            img.setAttribute('height', '15');
            img.setAttribute("visibility", "visible");
            img.appendChild(use);
            
            return img
        }
```

The obvious problem is that it is creating one overlay per position so that loading several thousand 1) takes time, 2) freezes the display, 3) makes the whole visualisation slow.
I wanted to group (< g >?) the identical symbol in one overlay and then adding only this overlay:
```
for (var idxSymbol in annotation_data.sets){
             svgImg=createSvgImg(symbolName);
             for (var idxPoint in annotation_data.sets[idxSymbol].points){
                         svgUse=addSvgSymbol(symbol, annotation_data.sets[idxSignal].points[idxPoint].x, annotation_data.sets[idxSignal].points[idxPoint].y);  
                        svgImg.appendChild(svgUse)      
              }
              OSViewer.addOverlay({
                         id: svgImg,
                         x:0,
                         y:0,
                         placement: 'CENTER',
                         checkResize: false,
              });                                 
}

function createSvgImg(symbolName){
            var img = document.createElementNS("http://www.w3.org/2000/svg", "svg");
              
            img.setAttribute('class', symbolName);
            img.setAttribute('width', '100%');
            img.setAttribute('height', '100%'); 
            img.setAttribute("visibility", "visible");

            return img
}
function addSvgSymbol(symbolValue, xRef, yRef){
            var use = document.createElementNS('http://www.w3.org/2000/svg', 'use');
           
            var symbolType = (symbolValue.image.rect.y-1)%(tempSymb.length);
            var colorType = Math.floor((symbolValue.image.rect.y-1)/(tempSymb.length*10));
            var symb= "img/myicons.svg#".concat(tempSymb[symbolType]);

            use.setAttributeNS("http://www.w3.org/1999/xlink", "xlink:href", symb);
            use.setAttribute('style',"fill: transparent; stroke: ".concat(tempColor[colorType]).concat("; stroke-width:8%"));

            use.setAttribute('x', xRef); 
            use.setAttribute('y',yRef);
            
            return use
 }
```
The problem is that the positions are all messed up when I am zooming in/out and that I am not sure I am tackling this problem the correct way.

Any help will be appreciated.
Thanks,

**Comment by iangilman** [comment]
For a case like this I recommend using the SVG Overlay plugin: 

https://github.com/openseadragon/svg-overlay

It gives you a single SVG surface that you can attach whatever SVG elements you want to, and the whole thing zooms and pans in sync with the viewer. It should be much more performant. 

It's possible that even this is too slow for 20,000 items. Another option would be to use the Canvas overlay plugin: 

https://github.com/altert/OpenSeadragonCanvasOverlay

Canvas is quite fast, and also you only have to draw the bits that make sense at the current zoom / pan position. 

You could also consider culling out unnecessary SVG elements from the SVG Overlay as you zoom / pan... I'm not sure if that would help. 

Anyway, there are some ideas to pursue!

**Comment by maxBombrun** [comment]
Hi,

Today I tried the svg-overlay and the canvasOverlay.
For the canvas, it tried to handle the svg symbol through an array, so that the svg is created in the for loops and then called onRedraw defined as such:
```
        var svg2Canvas= OSViewer.canvasOverlay({
            onRedraw:function() {
                for(var i = 0; i < overlayArray2canvas.length; i++){
                    var img = new Image();
                    img.src = "data:image/svg+xml;charset=utf-8,"+overlayArray2canvas[i];
                    svg2Canvas.context2d().drawImage(img,0,0); 
                }                    
            },
            clearBeforeRedraw:true
        } );
```
I think my conversion to image is not correct but I could not find a way to handle the definition of my symbol:
`<svg class="class1" width="100%" height="100%" visibility="visible"><use xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="img/myicons.svg#pentagram" style="fill: transparent; stroke: red; stroke-width:8%" width="20" height="20" x="100" y="100"></use></svg>`

For the svg-overlay, kind of the same problem but I fix it. I managed to add the symbol on the image but when I zoom in and out, the group g is translating and scaling but the symbol itself is not (probably because of the use). I write a OSViewer.addHandler('zoom', function(event)) function to resize the width and height in use and it is working. 
You are right, Ian, it is still taking some time when there are more than 10000 symbols. An approach display only the "visible" may be the solution.

Thanks,

**Comment by iangilman** [comment]
Interesting... can you pull the pentagram out of myicons.svg so you don't have that xlink? That should improve things on both the canvas and SVG side. Otherwise your SVG to canvas conversion looks reasonable (though for speed you'd want to build your images once at startup and just draw them during onRedraw). 

If you can get Canvas to work, it'll be a lot easier to draw only the "visible" items (and potentially draw placeholders or entirely omit when zoomed out far enough) with it.

**Comment by maxBombrun** [comment]
I managed to get the canvas work but I have some questions:
1) Is the position of the canvas is then relative to the viewport? One symbol on the top left of the canvas will stay on the top left of the view when I zoom and not get out of the field of view.
2) Would it be somehow possible to place all the symbols for one class in a svg and then, convert the svg in image -> canvas?

Thanks,

**Comment by iangilman** [comment]
1. Are you using the canvas plugin? I believe it modifies the canvas transform to match zooming and panning. Are you saying you want something that doesn't respond to zooming and panning? If so, you can draw directly on the viewer's canvas (after it draws each frame). 
2. I imagine you could have all of your symbols in a single SVG which you draw into an offscreen canvas, and then you draw portions of that canvas into the OSD canvas.

**Comment by maxBombrun** [comment]
Hi,

I had time to work a bit on this problem during summer but still not a perfect solution (even though I am getting close).
Here are my thoughts:
- I manage to display only circles (pie chart for example), at different locations in the viewer. To do so, I used the "OpenSeadragonCanvasOverlay"

```
cellOverlay = OSViewer.canvasOverlay({
          onRedraw:function() {
                    var viewPortBounds= OSViewer.viewport.viewportToImageRectangle(OSViewer.viewport.getBounds(true));
                    var defRadius=30;
                    for (var idxCell=0;idxCell<cellLocation.length; idxCell++) {
                              if (viewPortBounds.y<= cellLocation[idxCell].y && viewPortBounds.y+viewPortBounds.height >= cellLocation[idxCell].y && viewPortBounds.x <= cellLocation[idxCell].x && viewPortBounds.x +viewPortBounds.width >= cellLocation[idxCell].x) {
                                        cellOverlay.context2d().fillStyle = 'red';
                                        cellOverlay.context2d().beginPath();
                                        cellOverlay.context2d().arc(cellLocation[idxCell].x, cellLocation[idxCell].y,defRadius, 0, Math.PI * 2));
                                        cellOverlay.context2d().fill();
                              }
                    }
          },
          clearBeforeRedraw:true,
          setId: 'cellOverlayID'
});
OSViewer.raiseEvent('update-viewport');
```

It works fine with the zoom and I display only the circle visible in the viewPortBounds.

However, now I want to display different symbols (and still a lot), I followed Ian's advice and transfer my svgPaths in svg inline.
It kind of work if I use:
```
markerOverlay[symbolName] = markerOverlay[symbolName]  || {name: symbolName, canv: OSViewer.canvasOverlay({
          onRedraw:function() {
                    var viewPortBounds= OSViewer.viewport.viewportToImageRectangle(OSViewer.viewport.getBounds(true));
                    var  svgImg='<svg xmlns="http://www.w3.org/2000/svg"> <polygon points="0 12 12 24 24 12 12 0" fill="rgb(255, 0, 0)" stroke-width="1" stroke="rgb(0, 0, 0)"/> </svg>'
                    var imgMarker = new Image();
                    imgMarker.src="data:image/svg+xml;charset=utf-8," + svgImg;
                    for (var idxPoint=0;idxPoint<annotation_data.sets[idxSignal].points.length; idxPoint++) {
                              var signalLocation= new OpenSeadragon.Point(annotation_data.sets[idxSignal].points[idxPoint].x, annotation_data.sets[idxSignal].points[idxPoint].y);
                              if (viewPortBounds.y<= signalLocation.y && viewPortBounds.y+viewPortBounds.height >= signalLocation.y && viewPortBounds.x <= signalLocation.x && viewPortBounds.x +viewPortBounds.width >= signalLocation.x) {
                                        imgMarker.onload = function(){
                                                  markerOverlay[symbolName].canv.context2d().drawImage(imgMarker,signalLocation.x,signalLocation.y); 
                                        }
                              }
                    }
          },
          clearBeforeRedraw:true,
          setId: 'markerOverlayID_'+ symbolName
          })
}; 
OSViewer.raiseEvent('update-viewport'); 
```

but not if I used 
`var svgImg='<svg xmlns="http://www.w3.org/2000/svg" > <use xlink:href="#diamond"/>  </svg>';`

with diamond defined inline my html part:
```
<svg style="display: none;" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
          <symbol viewBox="0 0 100 100" id="diamond">
                    <polygon id="diamond" points="0 50 50 100 100 50 50 0"/>
          </symbol>
        
</svg>
```

**Comment by iangilman** [comment]
Glad you're making progress! I'm not surprised that the `xlink` version isn't working, since your image isn't actually really part of the page. I suppose you could try adding it to the page somewhere just to see if that fixed it. 

If you can't use xlink, can you still make it work in-line? Doesn't sound to me like too big of a problem, but maybe I don't understand the situation.

**Comment by maxBombrun** [comment]
Thanks @iangilman ,

I think it is my who lack knowledge about good behaviour with js/HTML.

I have a php file with an HTML part where my svg is defined as written in my previous post.
For me, this is creating the svg as part of my page.
Then, I have a js part which call it through the xlink.

I understand that xlink could not work in this case, I do not understand what you meant by "make it work in-line" in this case.

Thanks for helping.

**Comment by iangilman** [comment]
By "make it work inline" I just mean something like this:

```
var svgImg='<svg><symbol viewBox="0 0 100 100" id="diamond"><polygon id="diamond" points="0 50 50 100 100 50 50 0"/></symbol></svg>';
```

**Comment by maxBombrun** [comment]
Ah, yes sorry.
That was I meant in my previous post:

It kind of work if I use:
```
markerOverlay[symbolName] = markerOverlay[symbolName]  || {name: symbolName, canv: OSViewer.canvasOverlay({
          onRedraw:function() {
                    var viewPortBounds= OSViewer.viewport.viewportToImageRectangle(OSViewer.viewport.getBounds(true));
                    var  svgImg='<svg xmlns="http://www.w3.org/2000/svg"> <polygon points="0 12 12 24 24 12 12 0" fill="rgb(255, 0, 0)" stroke-width="1" stroke="rgb(0, 0, 0)"/> </svg>'
                    var imgMarker = new Image();
                    imgMarker.src="data:image/svg+xml;charset=utf-8," + svgImg;
                    for (var idxPoint=0;idxPoint<annotation_data.sets[idxSignal].points.length; idxPoint++) {
                              var signalLocation= new OpenSeadragon.Point(annotation_data.sets[idxSignal].points[idxPoint].x, annotation_data.sets[idxSignal].points[idxPoint].y);
                              if (viewPortBounds.y<= signalLocation.y && viewPortBounds.y+viewPortBounds.height >= signalLocation.y && viewPortBounds.x <= signalLocation.x && viewPortBounds.x +viewPortBounds.width >= signalLocation.x) {
                                        imgMarker.onload = function(){
                                                  markerOverlay[symbolName].canv.context2d().drawImage(imgMarker,signalLocation.x,signalLocation.y); 
                                        }
                              }
                    }
          },
          clearBeforeRedraw:true,
          setId: 'markerOverlayID_'+ symbolName
          })
}; 
OSViewer.raiseEvent('update-viewport'); 
```

So yes, I guess it is actually kind of solving my problem. I can create a function in the onRedraw which fetch the good svg definition. I will try that, thanks!

In this case, the problem comes from the signalLocation.x and .y .
My background image is originally 5226x3889 and my signal locations are defined based on this image, for example a.Point {x: 3653, y: 3801} but the canvas is defined otherwise (based on my screen definition??) i.e., <canvas id="markerOverlayID_ATB1" width="1453" height="406"></canvas>.
If I tried to OSViewer.viewport.pointFromPixel(currPoint); then the zoom changes their locations in the image.

**Comment by iangilman** [comment]
Converting from image coordinates to viewport coordinates would be:

```
OSViewer.viewport.imageToViewportCoordinates(currPoint);
```

**Comment by maxBombrun** [comment]
is CanvasOverlay working with the viewport?
Using OSViewer.viewport.imageToViewportCoordinates(currPoint); squeeze all my signal on the top left of my canvas since viewport ranges in [0,1].

**Comment by iangilman** [comment]
I believe Canvas overlay recently switched to using viewport coordinates... do you have the latest version?

**Comment by maxBombrun** [comment]
I use this one https://github.com/altert/OpenSeadragonCanvasOverlay on my server since June and the last commit is a year ago. I can still try to download it again.

**Comment by iangilman** [comment]
Actually, it looks like the Canvas overlay is using the image coordinates already, so maybe no conversion is necessary... have you tried that? 

The thing I was thinking of that got converted to viewport coordinates was @altert's Fabric overlay plugin (see https://github.com/altert/OpenseadragonFabricjsOverlay/issues/18). I've now proposed he do the same for the Canvas overlay (https://github.com/altert/OpenSeadragonCanvasOverlay/issues/5). That doesn't necessarily make a big difference for you, but for multi-image it does matter. 

At any rate, try it without conversion on the current version.

**Comment by maxBombrun** [comment]
Ok, I will try to explain the different steps that bring me to my solution.
First the main problem came from the onLoad of the image that I was using inside my onRedraw, wrong move, the onRedraw should be inside the onLoad:
```
imgMarker.onload = function(){
        markerOverlay[symbolName] = markerOverlay[symbolName]  || {name: symbolName, canv: OSViewer.canvasOverlay({
                onRedraw:function() {
                        var viewPortBounds= OSViewer.viewport.viewportToImageRectangle(OSViewer.viewport.getBounds(true));
                        for (var idxPoint=0;idxPoint<annotation_data.sets[idxSignal].points.length; idxPoint++) {
                                        var signalLocation= new OpenSeadragon.Point(annotation_data.sets[idxSignal].points[idxPoint].x,annotation_data.sets[idxSignal].points[idxPoint].y);
                                        if (viewPortBounds.y<= signalLocation.y && viewPortBounds.y+viewPortBounds.height >= signalLocation.y && viewPortBounds.x <= signalLocation.x && viewPortBounds.x +viewPortBounds.width >= signalLocation.x) {
                                                markerOverlay[symbolName].canv.context2d().drawImage(imgMarker, signalLocation.x, signalLocation.y); 
                                        }
                        }
                 },
                clearBeforeRedraw:true,
                setId: 'markerOverlayID_'+ symbolName
        })
        };
}
```
That created the marker for one symbol.
The problem after that was to display two canvas (i.e., two symbols overlayed). Only one was in the architecture (the first one), because in the canvasOverlay, the condition 
```
if (this._canvasOverlayInfo) {
            return this._canvasOverlayInfo;
} 
this._canvasOverlayInfo = new Overlay(this,options);
return this._canvasOverlayInfo;
```
was true -> no new Overlay. Therefore, I comment the if and then everything worked. Well, I have to had the option setId in the constructor so that:
```
this._canvas.id= options.setId; 
```
So every canvas has its own id and no problem occur to display several overlays.

The minor problem I have now, is that when I toggle on my signal, they appear on the left side of the image (background, openseadragon canvas) in the correct global shape. If I zoom in/zoom out, the handler "update-viewport" hits and they go on the image in the correct position.
```
this._viewer.addHandler('update-viewport', function() {
        self.resize();
        self._updateCanvas();
});
```
I tried to add:
OSViewer.raiseEvent('update-viewport');
after the function toggleSignals (which is actually needed to display the markers), but nothing change.

1) What do you think about my solution, 2) and about my new minor problem?
3) Is it not more power consuming to actually redraw all the markers all the time? Would it be smoother to just draw all of them, while zooming in/out will just hide the markers out of the container by itself? (I did not try yet, just wondering).

Thanks,

**Comment by iangilman** [comment]
Canvas by its very nature pretty much has to be a "redraw every frame" kind of thing, and fortunately it's optimized for that. If you want the "draw once and then transform as needed" approach, you should use the SVG overlay instead of the canvas overlay. 

Why are you using two canvases? Sounds like you have two modes you want to be able to switch between? If so, it would actually be more efficient to use a single canvas and just draw different things depending upon the mode (since you have to redraw every frame anyway).

**Comment by maxBombrun** [comment]
Sorry, I realise I did not reply to you.
Yes, I have one button for each symbol, when I click on one button it activates the symbol and several buttons can be activate in same time. I did not know about this "redraw every frame" so indeed I think a better version could draw only the symbol activated and in the viewport. 
Right now, since the idea is to have 4-5 symbols activated simultaneously it is not a problem but I will try to optimise that when I will have some free time.

I think we can say this topic is solved !

Thanks a lot for your help.

Cheers,
